index/types.go

```diff
@@ -21,22 +21,96 @@
 package index
 
 import (
-	"github.com/m3db/m3ninx/index/segment"
+	"regexp"
 
-	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/postings"
+
+	"github.com/m3db/m3x/errors"
 )
 
-// Index is a collection of segments.
+// Index is a collection of searchable documents.
 type Index interface {
-	segment.Readable
-	segment.Writable
+	// Insert inserts the given document into the index. The document is guaranteed to be
+	// searchable once the Insert method returns.
+	Insert(d doc.Document) error
+
+	// Snapshot returns a point-in-time snapshot of the index that provides a consistent
+	// view for searches.
+	Snapshot() (Snapshot, error)
+
+	// Seal marks the index as immutable. After Seal is called no more documents can be
+	// inserted into the index.
+	Seal() error
+
+	// Close closes the index and releases any internal resources.
+	Close() error
+}
+
+// Snapshot is a point-in-time snapshot of an index. It must be closed to
+// release its internal readers.
+type Snapshot interface {
+	// Readers returns the set of readers contained in the snapshot. Each call to Readers
+	// will return independent clones of the same Readers which the caller is responsible
+	// for closing.
+	Readers() Readers
+
+	// Close closes the snapshot and releases any internal resources.
+	Close() error
+}
+
+// Reader provides a point-in-time accessor to the documents in an index.
+type Reader interface {
+	// MatchTerm returns a postings list over all documents which match the given term.
+	MatchTerm(field, term []byte) (postings.List, error)
+
+	// MatchRegexp returns a postings list over all documents which match the given
+	// regular expression.
+	MatchRegexp(field, regexp []byte, compiled *regexp.Regexp) (postings.List, error)
+
+	// Docs returns an iterator over the documents whose IDs are in the provided
+	//  postings list.
+	Docs(pl postings.List) (doc.Iterator, error)
+
+	// Clone returns an independent clone of this Reader. The returned reader must be
+	// closed independently of this Reader.
+	Clone() (Reader, error)
```

Comment:
How is the `Clone()` method used?

Reply:
Talked offline, since a Reader maintains a reference count to the segment it was created from, `Clone` is used to increment the reference count and return a copy of the given Reader.